azkaban-common/src/main/java/azkaban/imagemgmt/daos/ImageVersionDao.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright 2020 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package azkaban.imagemgmt.daos;
+
+import azkaban.imagemgmt.exeception.ImageMgmtException;
+import azkaban.imagemgmt.models.ImageVersion;
+import azkaban.imagemgmt.dto.RequestContext;
+import java.util.List;
+
+/**
+ * Data access object (DAO) for accessing image version metadata. This interface defines method
+ * such as create, get image version metadata etc.
+ */
+public interface ImageVersionDao {
+  public int createImageVersion(ImageVersion imageVersion);
+  public List<ImageVersion> getImageVersion(RequestContext requestContext) throws ImageMgmtException;
```

Comment:
You should not pass request context to DAO. It should have specific parameters.

Reply:
RequestContext is a data transfer object and it just passes the data across layers. Internal builder is provided and hence the request builder can be build with required parameter. With this context, we can build RequestContext with the necessary parameters and pass the same to DAO methods specially for get/fetch/search etc. In this scenario need not pass multiple parameters and change the method signature if new parameters added in future.  In case of search with some optional parameters we can just check if the parameter is available or not and based on that for the query. Moreover, pagination parameters can also be passed across. This is specially useful for any search, get/fetch methods etc.

I think with the name getImageVersions the purpose of the method is not very clear. I agree if we have two mandatory parameters say imageType and imageVersion, we can have simply pass these two parameters in the DAO method. But the purpose here is not that. Let me refine the method name as findImageVersions as this is more of a search API. Below are the use cases it is trying to solve.

Given an imageType (mandatory parameter),
1. Give me all the versions.
2. Give me a specific version.
3. Give me all the active versions.
4. Give me all the unstable versions.
So, imageVersion and and state are optional parameters. Let me rename the method and include state in the API as well. This would make the purpose of this API more meaningful.

Reply:
Thank you for sharing the detail. It looks good in that case.